constellation-app/constellation
# Add blaze size preference #125

### Description of the Change
This pull request adds an extra UI tab to the preferences menu located in Setup -> Options -> Constellation -> Graph. The tab is responsible for global graph preferences upon loading a graph with no saved preferences. The issue was targeted at assigning a default blaze preference, but was extended to opacity also, due to ease of implementation and usability. Two slide controls have been added and control the values between 10 and 100 for both properties. 
The UI changes follow a similar layout structure as other preference tabs already utilized (Application). 
The previously assigned default values have been replaced with look-ups to globally stored preferences. In the case where no globally stored preference file exists, a default value still exists at the highest point in the lookup process (GraphPreferenceKeys).

The flow of logic is as follows:
 - Application start
 - Read of value from Preferences file (Default Value if no file exists)
 - Preferences Graph slider changed
 - Value written to Preferences file
 - New graph button clicked
 - New graph created with value stored in point 2
 
The behaviour above is not ideal, although allows the user to set a default value and have it refresh whenever the application restarts.



### Alternate Designs
The few options trialed to fix this issue were to utilize an accessor and mutator for fetching values from GraphPreferenceKeys, as opposed to requests to nbPreferences.forModule(...). Review of ways to "Refresh" or "Update" the SchemaFactory seemed not feasible, and were not implemented. Flush() - ing and Sync() - ing were both tested with the nbPreferences class in an attempt to ensure the values were written. The issue seemed to reflect the storage of the variable within application start and the lack of re-evaluation at run time. 


### Why Should This Be In Core?
This pull request adds functionality to allow a default value to be altered. Which, when creating a new graph, utilizes the new default value. It adds to the UX of the application and avoids making the user reset the blaze size whenever they create a new graph.


### Benefits
Extensibility of the UI for Graph property storage has been enhanced, which will support further development. The fix also, as mentioned above: reduces the amount of time that the user spends resetting blaze sizes when a new graph is made.


### Possible Drawbacks
Drawbacks of this implementation include the necessity to restart the application before changes become active. This is due to VisualConcept.java SchemaAttribute variables being assigned at application start. The current architecture does not allow for easy refactoring nor a proper workaround as many obstacles are in the way. Obstacles such as final declarations, unmodifiableCollections and statically typed classes make changes or updates to variables during run time difficult.


### Verification Process
All Tests were ran to verify regression, and human interaction with the tool was used as a basis for testing. 
 - launch constellation -> open new graph -> check blaze size = should be default value
 - launch constellation -> open preferences -> change preference values -> verify changed value in preferences.properties file = should be same value on slider
 - launch constellation -> change preferences -> open new graph = graph uses old preferences value (should use new preferences value, reason listed above)
 - launch constellation -> change preference values -> close application -> open application -> new graph -> check blaze values = should be saved new default value
 - launch constellation -> change blaze values in slider -> save file -> close file -> open file -> check blaze values = should be same as saved blaze size in file
 - launch constellation ->open multiple graphs -> change blaze slider on each -> verify they keep state and are different = states should be different



### Applicable Issues
As mentioned above, the minor issue of application restart needed for new preference value to become effective.
